stdlib/public/Darwin/Foundation/NSDictionary.swift

```diff
@@ -62,64 +73,100 @@ extension Dictionary : _ObjectiveCBridgeable {
   }
 
   public static func _forceBridgeFromObjectiveC(
-    _ d: NSDictionary,
+    _ x: NSDictionary,
     result: inout Dictionary?
   ) {
     if let native = [Key : Value]._bridgeFromObjectiveCAdoptingNativeStorageOf(
-        d as AnyObject) {
+        x as AnyObject) {
       result = native
       return
     }
 
     if _isBridgedVerbatimToObjectiveC(Key.self) &&
        _isBridgedVerbatimToObjectiveC(Value.self) {
-      result = [Key : Value](_cocoaDictionary: d)
-      return
-    }
-
-    if Key.self == String.self {
-      // String and NSString have different concepts of equality, so
-      // string-keyed NSDictionaries may generate key collisions when bridged
-      // over to Swift. See rdar://problem/35995647
-      var dict = Dictionary(minimumCapacity: d.count)
-      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-        let key = Swift._forceBridgeFromObjectiveC(
-          anyKey as AnyObject, Key.self)
-        let value = Swift._forceBridgeFromObjectiveC(
-          anyValue as AnyObject, Value.self)
-        // FIXME: Log a warning if `dict` already had a value for `key`
-        dict[key] = value
-      })
-      result = dict
+      result = [Key : Value](_cocoaDictionary: x)
       return
     }
 
     // `Dictionary<Key, Value>` where either `Key` or `Value` is a value type
     // may not be backed by an NSDictionary.
-    var builder = _DictionaryBuilder<Key, Value>(count: d.count)
-    d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
-      let anyObjectKey = anyKey as AnyObject
-      let anyObjectValue = anyValue as AnyObject
-      builder.add(
-          key: Swift._forceBridgeFromObjectiveC(anyObjectKey, Key.self),
-          value: Swift._forceBridgeFromObjectiveC(anyObjectValue, Value.self))
-    })
+    let fastSelf = unsafeBitCast(x, to: _NSDictionaryEnumerationHack.self)
+    var builder = _DictionaryBuilder<Key, Value>(count: x.count)
+    fastSelf.enumerateKeysAndObjects(options: 0) { (anyKey, anyValue, stopPtr) in
+      anyKey._withUnsafeGuaranteedRef { (anyKey) in
+        anyValue._withUnsafeGuaranteedRef { (anyValue) in
+          let key = Swift._forceBridgeFromObjectiveC(
+                  anyKey, Key.self)
+          let value = Swift._forceBridgeFromObjectiveC(
+                  anyValue, Value.self)
+          // String and NSString have different concepts of equality, so
+          // string-keyed NSDictionaries may generate key collisions when bridged
+          // over to Swift. See rdar://problem/35995647
+          if Key.self == String.self &&
+                !(unsafeBitCast(key, to: String.self)._isKnownNFC) {
```

Comment:
I don't think this is the right check. Just because *this* key is NFC doesn't mean a previous one wasn't.

Reply:
I also don't think it's a good idea to codify that there are no other bridged types that behave differently between Objective-C and Swift.

Reply:
Ok. I wouldn't bother reviewing it right now, I just wanted to try stuff out (in this case including Karoy's suggestion). I will take this into account if I get something useful though